### DISCUSSION TITLE
roadmapping iroh mobile libraries

### DISCUSSION THREAD

**Discussion by charroch** [n0-computer/iroh#517] [General]
Do you have a roadmap for the mobile libraries?

Are you planning rewrites in the platforms' specific languages (i.e. Kotlin/Swift) or a port of Iroh via cross-compilation?

I am currently looking at solutions for an upcoming application, and I have explored go-mobile, IPFS Lite, Rust IPFS (until discontinued). I was planning to investigate iroh and was hoping to understand better your roadmap.

Thanks,
Carl

**Comment by dignifiedquire** [comment]
> Do you have a roadmap for the mobile libraries?

[I will let @b5 reply to the roadmap part]

> Are you planning rewrites in the platforms' specific languages (i.e. Kotlin/Swift) or a port of Iroh via cross-compilation?

We are not planning any rewrites, the core code will stay in rust and be cross compiled for the target platform. Luckily both Java & Swift (and I think Kotlin, though no experience from me there) have great support for FFI bindings to C, so the iroh libraries can than be exposed using a C based api to those.

**Comment by b5** [comment]
Hey Carl, hope you don't mind me converting this to a discussion. Feels more appropriate to kick off a conversation about who needs what, and in what order.

to make sure we're all on the same page, when we talk about _iroh mobile_ we have a little more clarity in terms of definition, iroh mobile is _an importable library that provides IPFS functionality in a native app_. This sounds obvious, but is different from projects like iroh-share, which are about solving a specific problem using IPFS-type solutions. So the thing we're roadmapping here is "when can I have IPFS in my mobile app?".  Hopefully that's what you're after 😄.

### Plan:
Here's how we plan to get there:

1. Ship a version of iroh that splits storage, p2p, and gateway concerns into separate services ("iroh cloud")
1. Combine all three of those services into a single process, talking to each other over a memory channel ("iroh one")
1. Refactor iroh-one into a configurable library that lets you embed iroh into other rust applications ("iroh-embed")
1. Compose iroh-embed into a crate that exposes c bindings
1. Write wrapper libraries in Swift, Kotlin, Java, and Objective-C that both configure iroh-embed, and expose native language APIs

We're doing it this way to get the most mileage out of our code possible. Walking the cascade of cloud > laptop > rust library > mobile library means we'll end with a robust, spec compliant codebase that captures the various concerns of different platforms & deployment targets.

### Time
Right now we're on step three of the above. We'll be constantly refining & testing all parts of this. We haven't started on steps 4 & 5, and don't intend to until Q2 2023. I know this sounds like a long ways off, but we have a _lot_ of work on the internals to accomplish first.

Steps 4 & 5 have a smaller surface area than prior steps, but we haven't hired the folks that are actually doing that work yet. These things take time because in other areas we're hard at work designing entirely new data transfer protocols to get IPFS to fit on a phone.

**So, bottom line, I think the earliest you can expect to see an importable Kotlin, Java, Swift, or Objective-C library for iroh is June 2023.** Unless of course, members of the open source community want to jump in, but it's our expectation that we will build, ship & maintain these bits with number 0 staff time.

We do have at least one project building on Iroh that has _already_ deployed to mobile: but they've been at this for some time, and have experienced rust engineers on staff: https://github.com/capyloon/ So it's doable, just a very steep learning curve. Our job is to bring that curve down.

Hopefully this helps level set expectations around mobile. I'd welcome any feedback & would be happy to answer any questions.

**Reply by charroch** [reply]
Thank you for the write-up and share. That makes a lot of sense and Q2 feels sensible.

I am new to Rust but would not mind exploring 4. and 5. when iroh-embed is documented. If I get it to work within another Rust app, I could try to get it to work on Android and iOS.

**Reply by dignifiedquire** [reply]
This needs updating, but can serve as a first point of reference on what to do to get it working on mobile: https://github.com/deltachat/deltachat-core-rust/pull/3489

**Reply by MinmoTech** [reply]
I was wondering if you have considered kotlin multiplatform for your mobile wrapper, it might save a lot of duplicate work :)

**Reply by b5** [reply]
I doubt we'll be able to get away with such an abstraction, given how directly we'd expect folks to be interacting with the exported, language native APIs. The whole point of each library is to provide a native-language feel around a rust core

**Comment by abueide** [comment]
@b5 @dignifiedquire any updates on the mobile side of things? I'm a kotlin multiplatform developer. Currently I'm building a media library app targeting desktop, web, and ios/android that lets users sync/stream their library to any of their devices no matter what network they're on using the most direct link available. I've done some reading through the docs and Iroh seems like it could quite possibly be the perfect fit with the document system where a user could add all of their devices as an author and sync various metadata between their devices (e.g. number of plays on a song, if the user found a new song they liked on their phone, they just add it and it gets synced/streamed to the rest of their devices, and if one of the user's devices breaks or is lost the rest of the devices will have parity to restore the library).

Here's some options I've evaluated

libzt - has jvm, android, ios, but missing js/wasm targets.
tsnet - seems like a really good option I like tailscale a lot, just not a go engineer so figuring out how to create bindings for wasm/js, and jvm and objective c for ios seems like a pretty big hurdle.

I've also been itching to put some more effort into learning rust and help contribute to some rust projects and am trying to figure out if iroh will be the right fit to start investing some time into. I saw on the docs there is a "coming soon" tag for kotlin and that got me interested. Are you guys planning on providing a full kotlin multiplatform library that targets jvm, js/wasm, native & android environments or are you guys planning to just make jvm bindings available? I'd love to help out in this area, just looking for some direction to get started.

**Reply by abueide** [reply]
also cool thing is if we go for a full kotlin multiplatform wrapper it can export to javascript, typescript, jvm, swift & native versions of the library pretty easily and the interop is pretty good, that is one of the things jetbrains is investing in pretty heavily. Then we'd have a centralized place with a robust toolchain & test suite to target all of these platforms natively.

**Reply by Winterhuman** [reply]
They do plan on adding Kotlin bindings at some point in the future as mentioned at https://github.com/n0-computer/iroh-ffi

**Reply by dignifiedquire** [reply]
yes, kotlin bindings are planned as the next bindings

**Reply by b5** [reply]
Super exciting @abueide, your project is _exactly_ the kind of use case we're building iroh for. As dig has already mentioned, we're planning Kotlin in the near future, It'll land in https://github.com/n0-computer/iroh-ffi. I'll try to make sure we put some time into generating a base kotlin lib in the next month. It would be _awesome_ to have you test drive & give feedback.

As for targeting the web, we do have a plan, and will compile iroh for WASM. We're going to ship support for swift, kotlin, python, and go _first_, then move on to WASM, (nodejs & the browser). The browser requires a different approach on the networking stack. We have a plan there, it's basically UDP over websockets, but need time to write it out & get it working.

In the meantime, let's start figuring out what requirements you have for iroh! Mind getting into the details I'd also be happy to set up a vidchat. feel free to book me [here](https://cal.com/bfive/30min). I'd love to hear more about your project & how we can help.

**Reply by dignifiedquire** [reply]
I have a basic PR up, would love any help from folks who actually use kotlin in making it better & testing it: https://github.com/n0-computer/iroh-ffi/pull/10

**Reply by kusius** [reply]
Have also been following Iroh and waiting for the Kotlin bindings. Exciting news! Looking forward to check it out, testing it and contributing in any way I can too !
